distributed_shampoo/shampoo_utils.py

```diff
@@ -124,6 +126,69 @@ def multi_dim_cat(split_tensors: List[Tensor], num_splits: List[int]) -> Tensor:
     return merged_tensor[0]
 
 
+def convex_split(
+    tensor: Tensor, orig_shape: torch.Size, start_idx: int, end_idx: int
+) -> List[Tensor]:
+    """Chunks tensor across multiple dimensions to be convex.
+
+    Args:
+        tensor (Tensor): Flattened gradient or tensor to split.
+        orig_shape (torch.Size): Shape of original tensor that tensor is a slice of.
+        start_idx (int): Flattened index in original tensor where tensor starts.
+        end_idx (int): Flattened index in original tensor where tensor ends (inclusive).
+
+    Returns:
+        split_tensors (List[Tensor]): List of tensors.
+
+    """
+    if len(tensor.size()) != 1:
+        logger.info(
+            f"Input tensor is not flat, has shape {tensor.size()}. Continuing without splitting."
+        )
+    assert (
+        end_idx - start_idx + 1 == tensor.size()[0]
+    ), f"Start/end indices do not match tensor size: start {start_idx} end {end_idx}, tensor size {tensor.size()}"
+
+    split_tensors = []  # TODO: figure out if order matters
+    cumul_start_idx = start_idx
+    cumul_end_idx = end_idx
+    remainder = tensor
+    for i in range(len(orig_shape) - 1, -1, -1):
+        dim = orig_shape[i]
+        start = np.unravel_index(cumul_start_idx, orig_shape)
+        end = np.unravel_index(cumul_end_idx, orig_shape)
+        sidx = start[i]
+        eidx = end[i]
+
+        # final dimension case
+        if i == 0 or start[:i] == end[:i]:
+            shape = [-1] + list(orig_shape[i + 1 :])
+            split_tensors.append(remainder.view(shape))
+            break
+
+        if sidx % dim != 0:
+            shape = [dim - sidx] + list(orig_shape[i + 1 :])
+            size = prod(shape)
+            split_tensors.append(torch.narrow(remainder, 0, 0, size).view(shape))
+            remainder = torch.narrow(remainder, 0, size, len(remainder) - size)
+            cumul_start_idx += size
+
+        if remainder.numel() == 0:
+            break
+
+        if eidx != dim - 1:
+            shape = [eidx + 1] + list(orig_shape[i + 1 :])
+            size = prod(shape)
+            split_tensors.append(torch.narrow(remainder, 0, -size, size).view(shape))
+            remainder = torch.narrow(remainder, 0, 0, len(remainder) - size)
+            cumul_end_idx -= size
+
+        if remainder.numel() == 0:
+            break
```

Comment:
Let's discuss this code snippet on Wednesday? I wonder if there's a way to simplify this code and optimize this without calling `torch.narrow`. Let me think about this.